2_add-two-numbers.md

```diff
@@ -0,0 +1,145 @@
+ * struct ListNode {
+ *     int val;
+ *     ListNode *next;
+ *     ListNode() : val(0), next(nullptr) {}
+ *     ListNode(int x) : val(x), next(nullptr) {}
+ *     ListNode(int x, ListNode *next) : val(x), next(next) {}
+ * };
+ */
+class Solution {
+public:
+    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
+        ListNode* dummy_head = new ListNode(0);
+        ListNode* added = dummy_head;
+        int curry = 0;
+        while (l1 || l2 || curry) {
+            const int val = getValueOrDefault(l1) + getValueOrDefault(l2) + curry;
+            const int digit = val % 10;
+            curry = val / 10;
+            added->next = new ListNode(digit);
+            added = added->next;
+            l1 = getNextIfExists(l1);
+            l2 = getNextIfExists(l2);
+        }
+        return dummy_head->next;
+    }
+
+private:
+    static int getValueOrDefault(const ListNode* const node) {
+        if (!node) return 0;
+        return node->val;
+    }
+
+    static ListNode* getNextIfExists(const ListNode* const node) {
+        if (!node) return nullptr;
+        return node->next;
+    }
+};
+```
+
+## 2nd
+
+### 参考
+
+- https://discord.com/channels/1084280443945353267/1231966485610758196/1237404231225577592
+- https://discord.com/channels/1084280443945353267/1225849404037009609/1233844967777243318
+- https://discord.com/channels/1084280443945353267/1233603535862628432/1233768573718106253
+- https://discord.com/channels/1084280443945353267/1200089668901937312/1206949445359640576
+
+再帰バージョンで書いた。
+
+所要時間: 10:12
+
+n: l1, l2それぞれのノード数の大きい方
+- 時間計算量: O(n)
+- 空間計算量: O(n)
+
+リストのサイズが高々100なのでスタックオーバーフローすることはないだろう。
+スタックサイズ8MBとして、スタックフレーム辺りの引数、ローカル変数でポインタ3つ、intが2つ + ベースポインタとリターンアドレスで48byte+αくらいだろうか？とすると 8*2**20/48 ≒ 170K回くらいは深くできそう。
+(ポインタやレジスタのサイズを64bitとして計算している)
+
+```cpp
+class Solution {
+public:
+    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
+        return addTwoNumbersHelper(l1, l2, 0);
+    }
+
+private:
+    static ListNode* addTwoNumbersHelper(
+        const ListNode* const digit_node1,
+        const ListNode* const digit_node2,
+        const int curry
+    ) {
+        if (!digit_node1 && !digit_node2 && !curry) return nullptr;
+        ListNode* const digit_node = new ListNode(0);
+        const int total = getValOrDefault(digit_node1) + getValOrDefault(digit_node2) + curry;
+        digit_node->val = total % 10; // digit
+        digit_node->next = addTwoNumbersHelper(getNextIfExists(digit_node1), getNextIfExists(digit_node2), total / 10);
+        return digit_node;
+    }
+
+    static int getValOrDefault(const ListNode* const node) {
+        if (!node) return 0;
+        return node->val;
+    }
+
+    static ListNode* getNextIfExists(const ListNode* const node) {
+        if (!node) return nullptr;
+        return node->next;
+    }
+};
+```
+
+
+## 3rd
+
+
+```cpp
+class Solution {
+public:
+    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
+        ListNode dummy_head = ListNode();
```

Comment:
`ListNode dummy_head;` のほうがシンプルだと思います。

Reply:
https://en.cppreference.com/w/cpp/language/default_initialization
デフォルト初期化、あまり分かってなくて避けちゃうんですよね。

> if T is a (possibly cv-qualified) non-POD(until C++11) class type, the constructors are considered and subjected to [overload resolution](https://en.cppreference.com/w/cpp/language/overload_resolution) against the empty argument list. The constructor selected (which is one of the [default constructors](https://en.cppreference.com/w/cpp/language/default_constructor)) is called to provide the initial value for the new object;

ListNodeは non POD (`is_pod<ListNode>::value` が `false` でした) なのでデフォルトコンストラクタが呼ばれるってことなんですかね...ただC++11までと書いてあるしPODの条件も良くわからない。
今回はコピー初期化の書き方になってましたが、大体 `ListNode foo = {};` のリスト初期化を使います。こちらならどうでしょうか？

Reply:
「(until C++11)」の部分が何を表しているのか、自分も分かりませんでした。

`ListNode foo = {};` はあまり見ないように思います。

Google C++ Style Guide には、明示的な言及は見つけられませんでしたが、
`ListNode dummy_head;`
相当の書き方が例示されているようです。
https://google.github.io/styleguide/cppguide.html#Local_Variables

C++ 得意な方、補足をお願いできますでしょうか…。

Reply:
えー、私もこれは手を出したくないやつです。

C 言語の struct はメモリーの初期化がなされないので確保が速く、そこに memcpy したりします。C++ はこれを引き継いでいるわけですが、vector などでこれをやられたら大事件です。こういったものを区別して扱わないといけません。
C++11で、POD (Plain Old Data、C 言語由来) が trivial, trivially_copyable, standard_layout などに概念が整理されたためにその注釈がついていますね。実用上はあまり変わっていなさそうです。

あとは、未定義動作ですかね。プログラムは未定義動作が含まれていたらなんでもありなので、含まれないことを仮定して最適化していいという利点があり、その境界はきちんと決めないといけません。char はおそらく文字列として使われるので、未定義動作になりにくくなっているみたいですね。

リスト初期化はあまり見慣れていないのですが、これは C++11 からの機能で、私は03くらいまでの機能で書いていたというだけのことかと思います。

Reply:
うーんリファレンスやGoogle C++ Style Guide見てみましたが判断が難しい...

> リスト初期化はあまり見慣れていないのですが、これは C++11 からの機能で、私は03くらいまでの機能で書いていたというだけのことかと思います。

なるほどです。じゃあいったんリスト初期化でいこうかな、と思います。未初期化の変数関係の未定義動作も嫌ですし...何かご意見ありましたら後でも良いので是非。

Reply:
https://abseil.io/tips/146

Reply:
記事ありがとうございます。
ListNodeはUser-Provided Constructorsを定義しているので、default-initializationが実行されたときはそちらが呼ばれる。で、そのdefault constructorの定義は `ListNode() : val(0), next(nullptr) {}` なので、data memberが不定になることはない。
...という感じですかね。この挙動に毎回思いを馳せるのは辛いので、記事にも

> As these rules are not intuitive, the easiest rule to remember to ensure an object is initialized is to provide an initializer. This is called value-initialization and is distinct from default-initialization, which is what the compiler will perform otherwise for scalar and aggregate types.

と書いてるので明示的に初期化子をつけようと思います。あと

```cpp
struct Foo {
  int v = 0;
};
```

のようなdefault member initializationも良さそうですね (シンプルな構造体だったらこっちかなぁ...)。